From 4e0faafcddbfafa879e1efcda56115c3a1d15695 Mon Sep 17 00:00:00 2001
From: Darren Etheridge <detheridge@ti.com>
Date: Wed, 29 Nov 2023 10:42:31 -0600
Subject: [PATCH 1/2] services: memory coherency fix for HWR on BXS devices

Vendor supplied fix for memory coherency issues on BXS devices with
complex buses hierarchies.

This corresponds to:  PP137446_CL6514287.patch

Signed-off-by: Darren Etheridge <detheridge@ti.com>
---
 services/server/common/physmem.c              | 19 +++++++++++++++++++
 .../server/env/linux/physmem_osmem_linux.c    |  2 ++
 2 files changed, 21 insertions(+)

diff --git a/services/server/common/physmem.c b/services/server/common/physmem.c
index 138348a..4b5896e 100644
--- a/services/server/common/physmem.c
+++ b/services/server/common/physmem.c
@@ -416,6 +416,23 @@ static PVRSRV_ERROR _DevPhysHeapFromFlags(PVRSRV_MEMALLOCFLAGS_T uiFlags,
 	return PVRSRV_OK;
 }
 
+static INLINE void _AdjustCPUCacheFlags(PVRSRV_MEMALLOCFLAGS_T *puiFlags)
+{
+	if ((*puiFlags & (PVRSRV_MEMALLOCFLAG_CPU_READABLE |
+	                  PVRSRV_MEMALLOCFLAG_CPU_WRITEABLE |
+	                  PVRSRV_MEMALLOCFLAG_KERNEL_CPU_MAPPABLE)) == 0)
+	{
+		/* We don't need to upgrade if we don't map into the CPU */
+		return;
+	}
+
+	/* Clear the existing CPU cache flags */
+	*puiFlags &= ~(PVRSRV_MEMALLOCFLAG_CPU_CACHE_MODE_MASK);
+
+	/* Add CPU cached flags */
+	*puiFlags |= PVRSRV_MEMALLOCFLAG_CPU_CACHE_INCOHERENT;
+}
+
 PVRSRV_ERROR
 PhysmemNewRamBackedPMR_direct(
 	CONNECTION_DATA *psConnection, PVRSRV_DEVICE_NODE *psDevNode,
@@ -442,6 +459,8 @@ PhysmemNewRamBackedPMR_direct(
 	 */
 	PVR_UNREFERENCED_PARAMETER(uiAnnotationLength);
 
+	_AdjustCPUCacheFlags(&uiPMRFlags);
+
 	eError = _ValidateParams(ui32NumPhysChunks, ui32NumVirtChunks, uiFlags,
 				 &uiLog2AllocPageSize, &uiSize);
 	PVR_RETURN_IF_ERROR(eError);
diff --git a/services/server/env/linux/physmem_osmem_linux.c b/services/server/env/linux/physmem_osmem_linux.c
index 28af469..fd516c7 100644
--- a/services/server/env/linux/physmem_osmem_linux.c
+++ b/services/server/env/linux/physmem_osmem_linux.c
@@ -3624,6 +3624,8 @@ PhysmemNewOSRamBackedPMR(PHYS_HEAP *psPhysHeap, CONNECTION_DATA *psConnection,
 		goto errorOnParam;
 	}
 
+	PVR_ASSERT(ui32CPUCacheFlags == PVRSRV_MEMALLOCFLAG_CPU_CACHED);
+
 	if (PVRSRV_CHECK_CPU_CACHE_CLEAN(uiFlags)) {
 		ui32CPUCacheFlags |= PVRSRV_MEMALLOCFLAG_CPU_CACHE_CLEAN;
 	}
-- 
2.34.1

